src/coreComponents/constitutive/fluid/MultiFluidBase.cpp

```diff
@@ -70,6 +70,16 @@ MultiFluidBase::MultiFluidBase( string const & name, Group * const parent )
   registerExtrinsicData( extrinsicMeshData::multifluid::dPhaseViscosity_dTemperature{}, &m_phaseViscosity.dTemp );
   registerExtrinsicData( extrinsicMeshData::multifluid::dPhaseViscosity_dGlobalCompFraction{}, &m_phaseViscosity.dComp );
 
+  registerExtrinsicData( extrinsicMeshData::multifluid::phaseEnthalpy{}, &m_phaseEnthalpy.value );
+  registerExtrinsicData( extrinsicMeshData::multifluid::dPhaseEnthalpy_dPressure{}, &m_phaseEnthalpy.dPres );
+  registerExtrinsicData( extrinsicMeshData::multifluid::dPhaseEnthalpy_dTemperature{}, &m_phaseEnthalpy.dTemp );
+  registerExtrinsicData( extrinsicMeshData::multifluid::dPhaseEnthalpy_dGlobalCompFraction{}, &m_phaseEnthalpy.dComp );
+
+  registerExtrinsicData( extrinsicMeshData::multifluid::phaseInternalEnergy{}, &m_phaseInternalEnergy.value );
+  registerExtrinsicData( extrinsicMeshData::multifluid::dPhaseInternalEnergy_dPressure{}, &m_phaseInternalEnergy.dPres );
+  registerExtrinsicData( extrinsicMeshData::multifluid::dPhaseInternalEnergy_dTemperature{}, &m_phaseInternalEnergy.dTemp );
+  registerExtrinsicData( extrinsicMeshData::multifluid::dPhaseInternalEnergy_dGlobalCompFraction{}, &m_phaseInternalEnergy.dComp );
```

Comment:
@klevzoff @rrsettgast 
In this PR, I added some vectors in `MultiFluidBase.hpp`  (8 `arraysViews` for phase enthalpy and phase internal energy) and I ended up with a [problem that we had some time ago](https://github.com/GEOSX/GEOSX/pull/1325#discussion_r577888918) with the CO2-Brine fluid model: I cannot compile using CUDA for some compilers (it is fine on Lassen for some reason) because of a "formal parameter space overflow" in the fluid update. 

For example [here](https://app.travis-ci.com/github/GEOSX/GEOSX/jobs/554920358), I got:
```
/opt/GEOSX_TPL/raja/include/RAJA/policy/cuda/forall.hpp(137): Error: Formal parameter space overflowed (4392 bytes required, max 4096 bytes allowed) in function _ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256ENS_9Iterators16numeric_iteratorIllPlEEZN5geosx9PVTDriver7runTestINS8_12constitutive29MultiPhaseMultiComponentFluidINSB_8PVTProps19EzrokhiBrineDensityENSD_21EzrokhiBrineViscosityENSD_20SpanWagnerCO2DensityENSD_20FenghourCO2ViscosityENSD_13CO2SolubilityEEEEEvRT_RN7LvArray9ArrayViewIdLi2ELi1ElNSM_10ChaiBufferEEEEUliE_lEEvT1_T0_T2_
```
In the case above, the error takes place in a unit test, but I also get it when the fluid update kernel is called from `CompositionalMultiphaseBaseKernels.hpp` on Pecan.

For this combination of CO2-brine models, we have to pass to the fluid update GPU kernel:
- 32 `arrayViews`
- 7 `TableFunction::KernelWrappers` (ultimately 9 when we do the enthalpy and internal energy updates)

Is there an approach that would allow us to overcome this issue? 
If the `TableFunction::KernelWrappers` are too large, should we go back to the (not very beautiful) `arrayView1d< TableFunction::KernelWrappers >` of size 1?

I am tagging @andrea-borio who had the same issue for the `MultiphasePoromechanicsKernel` with VEM.

Reply:
Thx @klevzoff 
 
> 1. Combine derivative arrays for `dP`, `dT`, `dC` into one array, having the same number of dimensions as `dC`, but with +2 in runtime size. [...] The downsides are slight reduction in readability and a minor performance regression on CPU in non-thermal cases since temperature values will be force-read into cache but remain unused (GPU should not be affected because due to different memory layout temperature values are effectively separated from the rest).

Is it possible, instead of providing a table, to provide a `class` holding this "_augmented_" array and providing accessors to the `dP`, `dT`, `dC` data? Something _like_
```c++
DT = 0; DP = 1; ...
double SomeClass::dT( int a, int b, int c ) const {
    this->array[ DT ][ a ][ b ][ c ];
}
double SomeClass::dP( int a, int b, int c ) const {
    this->array[ DP ][ a ][ b ][ c ];
}
```
so we can do `fields->dT( a, b, c )`? Would it still help about the memory issues we are speaking about?

In terms of readability, I'm not sure it's a real issue. In terms of memory layout performance, I'm sure we can `template meta-programming` something 🤣 

> 2. Move `TableFunction` wrappers into global memory (via `array1d` storage or similar), saving ~640 bytes (80 bytes per table). The downside are all the additional dynamic allocations than will need to happen on every kernel launch [...]

Are we (you 🤣 ) able to estimate the time of these additional allocation?
Is it possible, if we go this way, to implement dedicated "container" (instead of the `array1d` trick) so the intent would be clearer?

> 3. Create some sort of class that aggregates together multiple array views of equal size, saving on not storing repeated dims and strides. Depending on how we group arrays, we could save in the range of 256-1024 bytes (hard to estimate precisely). I'm struggling a bit to imagine what the interface and usage of that class would look like.

Do you have in mind something like I write in `1.` or is it something completely different?
Actually, would an extended version of https://github.com/GEOSX/GEOSX/pull/1750 (not serving the arrays, but accessing the data) be an example? If it's the case, would keeping up in https://github.com/GEOSX/GEOSX/pull/1750 's direction help at encapsulating, before we can change the contracts?

Reply:
Thanks very much for these suggestions @klevzoff and @TotoGaz, I am going to try Suggestion 1. (combining derivative arrays) in a separate PR so we can discuss something concrete soon.

Reply:
> Do you have in mind something like I write in `1.` or is it something completely different?

What I had in mind there is somewhat similar, but implemented in `LvArray` and thus devoid of any physical interpretation, e.g. (simplifying a bit)
```c++
template< int NUM_ARRAYS, typename T, int NDIM, typename INDEX_TYPE, typename PERM, typename BUFFER >
class MultiArrayView
{
public:
  template< typename ... INDICES >
  T & operator( int const array_index, INDICES ... indices )
  {
     return m_buffers[ array_index ][ linearIndex( indices ... ) ];
  }
private:
  INDEX_TYPE m_dims[NDIM];    // we store this once instead if NUM_ARRAYS times
  INDEX_TYPE m_strides[NDIM]; // ditto
  BUFFER m_buffers[NUM_ARRAYS];
};
...
// packs together views into all derivative arrays of matching size
MultiArrayView< 3, real64, 4, ... > dPhaseProps_dComp = ...
real64 const dPhaseDens_dC = dPhaseProps_dComp( 0, k, q, ip, ic );
real64 const dPhaseVisc_dC = dPhaseProps_dComp( 1, k, q, ip, ic );
real64 const dPhaseMassDens_dC = dPhaseProps_dComp( 2, k, q, ip, ic );
```
Of course, we can then add a GEOSX-side wrapper that gives "physical" names to those 0, 1, 2 indices.

Reply:
> Is it possible, instead of providing a table, to provide a `class` holding this "_augmented_" array and providing accessors to the `dP`, `dT`, `dC` data? Something _like_
> ...
> so we can do `fields->dT( a, b, c )`? Would it still help about the memory issues we are speaking about?

Yes, to be precise, it could look something like (here I also mashed values with derivatives, which may or may not be desired):
```c++
template< int NDIM, typename PERM >
class CompositionalVarView
{
public:
  template< typename ... INDICES >
  real64 & value( INDICES ... indices ) 
  { return m_view( indices..., 0 ); } // 0 is offset of value

  template< typename ... INDICES >
  real64 & dP( INDICES ... indices ) 
  { return m_view( indices..., 1 ); } // 1 is offset of pressure derivative

  template< typename ... INDICES >
  real64 & dT( INDICES ... indices ) 
  { return m_view( indices..., 2 ); } // 1 is offset of temperature derivative

  template< typename ... INDICES >
  real64 & dT( integer jc, INDICES ... indices ) // jc can't be after the pack, which is unfortunate
  { return m_view( indices..., 3 + jc ); } // 3 is offset of composition derivatives

private:
  // +1 accounts for extra derivative dimension
  ArrayView< real64, NDIM + 1, PERM > m_view;
}
...
CompositionalVarView< 3, multifluid::LAYOUT_PHASE > density = ...
forall( k, q, ip, jc )
{
  real64 const dens = density.value( k, q, ip );
  real64 const dDens_dP = density.dP( k, q, ip );
  real64 const dDens_dT = density.dT( k, q, ip );
  real64 const dDens_dC = density.dC( jc, k, q, ip ); // :sad_face:
}
```

Unlike the previous reply, this involves changing the actual storage of the data. But it results in more logical grouping IMO.